File: Python/structure_output.py
from pydantic import BaseModel, Field
from strands import Agent, tool
import re

class ResultadoConversion(BaseModel):
    monto_original: float
    moneda_original: str
    monto_final: float
    moneda_final: str
    taza_usada:float
    pasos: list[str] = Field(description="Explicación de cada paso del cálculo")


@tool
def calculadora(expresion: str) -> str:
    """Evalúa una expresión matemática simple, ej: '23 * 4 + 1'"""
    try:
        if not re.match(r"^[\d\s+\-*/().]+$", expresion): # filtro basico con regex para permitir solo números, operadores y paréntesis
            return "Error: la expresión contiene caracteres no permitidos."
        resultado = eval(expresion, {"__builtins__": {}})
        return f"El resultado es {resultado}"
    except ZeroDivisionError:
        return "Error: división entre cero."
    except Exception as e:
        return f"Error al evaluar la expresión: {e}"

@tool
def sumar(numeros: list[float]) -> str:
    """Suma n cantidad de números, ej: '1, 2, 3'"""
    if not isinstance(numeros, list):
        return "Error: se espera una lista de números."
    if not numeros:
        return "Error: no se proporcionaron números para sumar."
    if not all(isinstance(x, (int, float)) and not isinstance(x, bool) for x in numeros):
        return "Error: todos los elementos deben ser números válidos."
    resultado = sum(numeros)
    return f"El resultado de la suma es {resultado}"

@tool
def currency_converter(amount: float, fromCurrency: str, toCurrency: str) -> str:
    """Convierte MXN a USD y viceversa, ej: '100 USD a MXN'"""
    # Aquí podrías implementar la lógica de conversión de divisas usando una API externa
    # Por simplicidad, vamos a devolver un mensaje simulado
    try:
        if fromCurrency == toCurrency:
            return f"{amount} {fromCurrency} equivalen a {amount} {toCurrency}"

        tazas_en_mx = {"USD": 18, "MXN": 1}
        amount_in_mxn = amount * tazas_en_mx[fromCurrency]
        resultado = amount_in_mxn / tazas_en_mx[toCurrency]
        return f"Convertido {amount} {fromCurrency} a {toCurrency}: {resultado:.2f} {toCurrency}"
    except Exception as e:
        return f"Error al convertir la moneda: {e}"

agent = Agent(
    model="us.anthropic.claude-haiku-4-5-20251001-v1:0",
    system_prompt="Eres un asistente financiero. Usa las herramientas disponibles para calcular y convertir montos.",
    tools=[calculadora, sumar, currency_converter],
    structured_output_model=ResultadoConversion
)

respuesta = agent(
    "Convierte 100 MXN a USD, súmale 50.44 USD y conviértelo de vuelta a MXN",
)

print("Respuesta: ", respuesta)
print("Estructurada: ", respuesta.structured_output)
# Podemos manejar errores de salida estructurada si es necesario con StructuredOutputException
# El import es: from strands.types.exceptions import StructuredOutputException
# ResultadoConversion(monto_original=100, moneda_original='MXN', monto_final=1008.0, moneda_final='MXN', pasos=[...])

File: TypeScript/structureOutput.ts
import { Agent, tool, StructuredOutputError } from "@strands-agents/sdk";
import { z } from "zod";

// ...reusa calculadora y CurrencyConverter del Ejercicio 1
const calculadora = tool({
  name: "calculadora",
  description: "Evalúa una expresión matemática simple, ej: '23 * 4 + 1'",
  inputSchema: z.object({
    expresion: z.string(),
  }),
  callback: async ({ expresion }) => {
    try {
      if (!/^[\d\s+\-*/().]+$/.test(expresion)) {
        return `Error: la expresión contiene caracteres no permitidos.`;
      }
      const resultado = Function(`"use strict"; return (${expresion})`)();
      if (typeof resultado !== "number" || isNaN(resultado) || !isFinite(resultado)) {
        return `Error: la expresión no es válida, posible valor nulo o infinito.`;
      }
      return `El resultado es ${resultado}`;
    } catch (e) {
      return `Error al evaluar la expresión: ${e instanceof Error ? e.message : String(e)}`;
    }
  },
});

const sumar = tool({
  name: "Sumar_tool",
  description: "Suma n cantidad de números, ej: '1, 2, 3'",
  inputSchema: z.object({
    numeros: z.array(z.number()),
  }),
  callback: async ({ numeros }) => {
    try {
      if (!Array.isArray(numeros) || numeros.length === 0) {
        return "Error: no se proporcionaron números para sumar o la lista está vacía.";
      }
      if (!numeros.every(num => typeof num === "number" && !isNaN(num))) {
        return "Error: todos los elementos deben ser números válidos.";
      }
      const resultado = numeros.reduce((acc, curr) => acc + curr, 0);
      return `El resultado de la suma es ${resultado}`;
    } catch (error) {
      return `Error al sumar los números: ${error instanceof Error ? error.message : String(error)}`;
    }
  },
})

const CurrencyConverter = tool({
  name: "currency_converter",
  description: "Convierte una cantidad entre MXN y USD (tasa simulada: 1 USD = 18 MXN). Recibe amount, fromCurrency y toCurrency (usa 'MXN' o 'USD').",
  inputSchema: z.object({
    amount: z.number(),
    fromCurrency: z.enum(["MXN", "USD"]),
    toCurrency: z.enum(["MXN", "USD"]),
  }),
  callback: ({ amount, fromCurrency, toCurrency }) => {
    try {
      if (fromCurrency === toCurrency) {
      return `${amount} ${fromCurrency} equivalen a ${amount} ${toCurrency}`;
    }
    // Tasas relativas a MXN (moneda base)
    const tasasEnMxn: Record<string, number> = { MXN: 1, USD: 18 };
    const amountEnMxn = amount * tasasEnMxn[fromCurrency];
    const resultado = amountEnMxn / tasasEnMxn[toCurrency];
    return `${amount} ${fromCurrency} equivalen a ${resultado.toFixed(2)} ${toCurrency}`;
    } catch (error) {
      return `Error al convertir la moneda: ${error instanceof Error ? error.message : String(error)}`;
    }
  },
});

const resultadoSchema = z.object({
  montoOriginal: z.number(),
  monedaOriginal: z.string(),
  montoFinal: z.number(),
  monedaFinal: z.string(),
  tazaUsada: z.number(),
  pasos: z.array(z.string()).describe("Explicación de cada paso del cálculo"),
});

const agent = new Agent({
  model: "us.anthropic.claude-haiku-4-5-20251001-v1:0",
  systemPrompt: "Eres un asistente financiero. Usa las herramientas disponibles para calcular y convertir montos.",
  tools: [calculadora, sumar, CurrencyConverter],
});
try {
    const respuesta = await agent.invoke(
    "Convierte 100 MXN a USD, súmale 50.44 USD y conviértelo de vuelta a MXN",
    { structuredOutputSchema: resultadoSchema } // ajusta el nombre exacto de la opción si difiere en tu versión del SDK
    );

    console.log("Respuesta: ", respuesta);
    console.log("Estructurada: ", respuesta.structuredOutput);
    // { montoOriginal: 100, monedaOriginal: "MXN", montoFinal: 1008, monedaFinal: "MXN", pasos: [...] }
} catch (error) {
    // podemos capturar errores específicos de salida estructurada
    if (error instanceof StructuredOutputError) {
        console.error("Error al obtener la salida estructurada:", error.message);
    } else {
        console.error("Error inesperado:", error);
    }
}

